chainer/testing/function.py

```diff
@@ -0,0 +1,530 @@
+        Returns a tuple of the second order input gradient arrays of type
+        :class:`numpy.ndarray`.
+        ``input_template`` is a tuple of template arrays. The returned arrays
+        are expected to have the same shapes and dtypes as the template arrays.
+
+    .. rubric:: Class attributes
+
+    The base class can define the following class attributes to control the
+    behavior of the tests.
+
+    ``forward_test`` (bool):
+        Whether to test forward computation. ``True`` by default.
+
+    ``backward_test`` (bool):
+        Whether to test backward computation. ``True`` by default.
+
+    ``double_backward_test`` (bool):
+        Whether to test double-backward computation. ``True`` by default.
+
+    ``dodge_nondifferentiable`` (bool):
+        Enable non-differentiable point detection in numerical gradient
+        calculation. If the inputs returned by ``generate_inputs`` turns
+        out to be a non-differentiable point, the test will repeatedly resample
+        inputs until a differentiable point will be finally sampled.
+        ``False`` by default.
+
+    ``contiguous`` (None or 'C'):
+        Specifies the contiguousness of incoming arrays (i.e. inputs, output
+        gradients, and the second order input gradients). If ``None``, the
+        arrays will be non-contiguous as long as possible. If ``'C'``, the
+        arrays will be C-contiguous. ``None`` by default.
+
+    .. note::
+
+       This decorator assumes :func:`chainer.testing.inject_backend_tests`
+       is used together. See the example below.
+
+    .. admonition:: Example
+
+       .. testcode::
+
+          @chainer.testing.inject_backend_tests(
+              None,
+              [
+                  {},  # CPU
+                  {'use_cuda': True},  # GPU
+              ])
+          @chainer.testing.function_test()
+          class TestReLU(unittest.TestCase):
+
+              # ReLU function has a non-differentiable point around zero, so
+              # dodge_nondifferentiable should be set to True.
+              dodge_nondifferentiable = True
+
+              def generate_inputs(self):
+                  x = numpy.random.uniform(-1, 1, (2, 3)).astype(numpy.float32)
+                  return x,
+
+              def forward(self, inputs, device):
+                  x, = inputs
+                  return F.relu(x),
+
+              def forward_expected(self, inputs):
+                  x, = inputs
+                  expected = x.copy()
+                  expected[expected < 0] = 0
+                  return expected,
+    """
+
+    def wrap(cls):
+        # Assign class attributes
+        attrs = [
+            ('forward_test', True),
+            ('backward_test', True),
+            ('double_backward_test', True),
+            ('dodge_nondifferentiable', False),
+            ('contiguous', None),
+        ]
+        for attr_name, default_value in attrs:
+            if not hasattr(cls, attr_name):
```

Comment:
I think the interface would be cleaner passing these as arguments to `function_test` instead of setting them as class scope members (easier to follow documentation). Do you have comments regarding this?